### PR TITLE
feat: init indexer setup

### PR DESCRIPTION
**Pull request by Majorfi**
Init the Indexer system to be able to only get new elements

image

The indexer is started after the init and will live forever with try/catch fallbacks:
```
-> Index everything since inception
-> Start the indexer daemon
-> Live forever
-> Is killed because the node explode
-> Cannot connect to node for 60s with ws connection
-> Fallback to RPC fetch to get all events from last synced block to now
-> Oh, node un-exploded
-> Daemon resurected and ready
```

For now only listening to `newVaults` events for the registries V1 and V2

## How to test:
Start ganache: `ganache-cli -f yournodeuri`
Start yDaemon and wait for server ready
Start brownie: `brownie console --network mainnet-fork`
In brownie, run:
```py
me = yourAddress
token = "0xD533a949740bb3306d119CC777fa900bA034cd52"
registry = Contract.from_explorer("0x50c1a2eA0a861A967D9d0FFE2AE4012c2E053804")
registry.newVault(token, me, me, "TestCRV", "tCRV", {"from": "0xFEB4acf3df3cDEA7399794D0869ef76A6EfAff52"})
```

This should trigger a new vault detection

### CODE REVIEW

File: common/ethereum/utils.go [hunk 1]
```diff
@@ -58,3 +59,20 @@ func randomSigner() *bind.TransactOpts {
...
 }
+
+// GetWSClient returns the current ws connection for a specific chain
+func GetWSClient(chainID uint64) (*ethclient.Client, error) {
+	if WS[chainID] == nil {
+		uri := GetRPCURI(chainID)
+		uri = `http://localhost:8545`
+		if uri[:4] == `http` {
+			uri = `ws` + uri[4:]
+		}
```
[the remark this answers is not in the document]
Comment (mariuspod) [reply common/ethereum/utils.go:70]: can you use the standard golang URL package here instead?
https://pkg.go.dev/net/url#URL

File: internal/registries/fetcher.go [hunk 2]
```diff
@@ -192,8 +197,12 @@ func RetrieveAllVaults(
 	wg := &sync.WaitGroup{}
 	for _, registry := range env.YEARN_REGISTRIES[chainID] {
 		wg.Add(2)
-		go filterNewVaults(chainID, registry.Address, registry.Version, registry.Activation, &vaultsList, wg)
-		go filterNewExperimentalVault(chainID, registry.Address, registry.Version, registry.Activation, &vaultsListExperimental, wg)
+		startBlock := fromBlock
+		if fromBlock == 0 {
+			startBlock = registry.Activation
+		}
```
[the remark this answers is not in the document]
Comment (mariuspod) [reply internal/registries/fetcher.go:203]: I'd not use to semantically similar variables here `startBlock`, `fromBlock`. Instead maybe pass it via args or set a local default if it's not set.

File: common/helpers/helpers.go [hunk 3]
```diff
@@ -150,3 +152,34 @@ func AddressIsValid(address common.Address, chainID uint64) bool {
...
+		return false
+	}
+	return something[0].(bool)
+}
+func DecodeAddress(something []interface{}) ethcommon.Address {
+	if len(something) == 0 {
+		return ethcommon.Address{}
+	}
+	return something[0].(ethcommon.Address)
+}
```
[the remark this answers is not in the document]
Comment (mariuspod) [reply common/helpers/helpers.go:185]: maybe prefer `switch...case` here if possible

File: common/ethereum/utils.go [hunk 4]
```diff
@@ -58,3 +59,20 @@ func randomSigner() *bind.TransactOpts {
 	signer.GasPrice = big.NewInt(0)
 	return signer
 }
+
+// GetWSClient returns the current ws connection for a specific chain
+func GetWSClient(chainID uint64) (*ethclient.Client, error) {
+	if WS[chainID] == nil {
+		uri := GetRPCURI(chainID)
+		uri = `http://localhost:8545`
```
[the remark this answers is not in the document]
Comment (mariuspod) [reply common/ethereum/utils.go:67]: override via ENV variable and access via a helper function:
`uri := getEnv('FTM_RPC_URL', 'http://localhost:8545')`
```
func getEnv(key, fallback string) string {
    if value, ok := os.LookupEnv(key); ok {
        return value
    }
    return fallback
}
```
see: https://stackoverflow.com/a/40326580

**Review by kyhoon — DISMISSED** [review]

**Pull request merged**